src/util/TickMath.ts

```diff
@@ -183,4 +188,11 @@ export abstract class TickMath {
     }
     return msb;
   }
+
+  static tickSpacingToMaxLiquidityPerTick(tickSpacing: number): JSBI {
+    const minTick = (this.MIN_TICK / tickSpacing) * tickSpacing;
+    const maxTick = (this.MAX_TICK / tickSpacing) * tickSpacing;
+    const numTicks = (maxTick - minTick) / tickSpacing + 1;
+    return JSBI.divide(MaxUint128, JSBI.BigInt(numTicks));
+  }
```

Comment:
It's annoying to simulate maths from contract exactly because there is only integer at that side and operation on negative integers is kind of special from traditional language. That is, division as well as right shifts used rounding towards zero. So let's change to JSBI way which is doing good. I have tested the result with contract on mainnet.

Reply:
const minTick = JSBI.multiply(
        JSBI.divide(JSBI.BigInt(MIN_TICK), JSBI.BigInt(tickSpacing)),
        JSBI.BigInt(tickSpacing)
      );
      const maxTick = JSBI.multiply(
        JSBI.divide(JSBI.BigInt(MAX_TICK), JSBI.BigInt(tickSpacing)),
        JSBI.BigInt(tickSpacing)
      );
      const numTicks = JSBI.add(
        JSBI.divide(JSBI.subtract(maxTick, minTick), JSBI.BigInt(tickSpacing)),
        ONE
      );
      return JSBI.divide(MaxUint128, JSBI.BigInt(numTicks));